contrib/zenazn/goji/web/goji.go

```diff
@@ -0,0 +1,44 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+// Package web provides functions to trace the zenazn/goji/web package (https://github.com/zenazn/goji).
+package web // import "gopkg.in/DataDog/dd-trace-go.v1/contrib/zenazn/goji/web"
+
+import (
+	"fmt"
+	"math"
+	"net/http"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/contrib/internal/httputil"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+
+	"github.com/zenazn/goji/web"
+)
+
+// Middleware returns a goji middleware function that will trace incoming requests.
+// If goji's Router middleware is also installed, the tracer will be able to determine
+// the original route name (e.g. "/user/:id"), and include it as part of the traces' resource
+// names.
+func Middleware(opts ...Option) func(*web.C, http.Handler) http.Handler {
+	cfg := new(config)
+	defaults(cfg)
+	for _, fn := range opts {
+		fn(cfg)
+	}
+	if !math.IsNaN(cfg.analyticsRate) {
+		cfg.spanOpts = append(cfg.spanOpts, tracer.Tag(ext.EventSampleRate, cfg.analyticsRate))
+	}
+	return func(c *web.C, h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			resource := r.Method
+			p := web.GetMatch(*c).RawPattern()
+			if p != nil {
```

Comment:
This is quite unfortunate. Should we use `"unknown"` like we do in [other places](https://github.com/DataDog/dd-trace-go/blob/9f7532307476d362c068267dae6390c8320f1ff5/contrib/gorilla/mux/mux_test.go#L41) in an `else` branch?

Reply:
Yes, this is really unfortunate, especially since without the `Router` middleware installed, p will always be `nil` as far as I can tell. I read the documentation and played with goji and its middleware for a long time trying to find a way to reliably get the route and there doesn't seem to be one.

In an early iteration of the integration, I used `"unknown"` like we do in other places. However, I don't think it fits well here because you will often end up with routes that are *all* `"unknown"`. In the other integrations, `"unknown"` seems to be an uncommon exception, where it can distinguish itself from other traces with routes. Here, adding `"unknown"` adds nothing since they will often all be `"unknown"`, and it will end up just being text noise. With more complicated nested router setups, it's possible to have mixed routes and `"unknown"`, but I don't know how common that is, and I'm not sure missing the `"unknown"` detracts much from that case. 

I'm happy to add it in if you think it's important, but I think in this case it's better to leave it out.

Reply:
I don't think it's important. And I agree `unknown` could cause a lot of noise. It would be nice if people would have some way of getting notified that they need to set the router. They probably want to see the routes in their APM.

Can we add a `sync.Once` and `log.Warn("contrib/zenazn/goji: routes are unavailable. To enable them use the goji.DefaultMux.Router middleware.")` just so that the user gets at least 1 log message letting him know why he isn't seeing things (other than not having read the docs :) )